web/packages/teleterm/src/mainProcess/clusterStore/clusterStore.ts

```diff
@@ -0,0 +1,180 @@
+  async add(proxyAddress: string): Promise<Cluster> {
+    const { response } = await this.tshdClient.addCluster({
+      name: proxyAddress,
+    });
+
+    await this.update(draft => {
+      // Do not overwrite the existing cluster;
+      // otherwise we may lose properties fetched from the auth server.
+      // Consider separating properties read from profile and those
+      // fetched from the auth server at the RPC message level.
+      if (draft.has(response.uri)) {
+        return;
+      }
+      draft.set(response.uri, response);
+    });
+
+    return response;
+  }
+
+  /** Logs out of the cluster and removes its profile.*/
+  async logout(uri: RootClusterUri): Promise<void> {
+    // TODO(gzdunek): logout and removeCluster should be combined into
+    //  a single acton in tshd.
+    await this.tshdClient.logout({ clusterUri: uri });
+    await this.tshdClient.removeCluster({ clusterUri: uri });
+    await this.update(draft => {
+      for (let d of draft.values()) {
+        if (routing.belongsToProfile(uri, d.uri)) {
+          draft.delete(d.uri);
+        }
+      }
+    });
+  }
+
+  /**
+   * Synchronizes the root clusters.
+   * Does not make a network call, only reads profiles from the disk.
+   */
+  async syncRootClusters(): Promise<void> {
+    const { response } = await this.tshdClient.listRootClusters({});
+    await this.update(draft => {
+      draft.clear();
+      response.clusters.forEach(cluster => {
+        draft.set(cluster.uri, cluster);
+      });
+    });
+  }
+
+  /**
+   * Synchronizes the root cluster.
+   * Makes a network call to get cluster details and its leaf clusters.
+   */
+  async sync(uri: RootClusterUri): Promise<void> {
+    try {
+      const { cluster, leafs } = await getClusterAndLeafs(this.tshdClient, uri);
+
+      await this.update(draft => {
+        draft.set(cluster.uri, cluster);
+        leafs.forEach(leaf => {
+          draft.set(leaf.uri, leaf);
+        });
+      });
+    } catch (error) {
+      await this.update(draft => {
+        const cluster = draft.get(uri);
+        if (cluster) {
+          // TODO(gzdunek): We should rather store the cluster synchronization status,
+          // so the callsites could check it before reading the field.
+          // The workaround is to update the field in case of a failure,
+          // so the places that wait for showResources !== UNSPECIFIED don't get stuck indefinitely.
+          cluster.showResources = ShowResources.ACCESSIBLE_ONLY;
+        }
+      });
+      throw error;
+    }
+  }
+
+  getRootClusters(): Cluster[] {
+    return this.state
+      .values()
+      .toArray()
+      .filter(c => !c.leaf);
+  }
+
+  getState(): State {
+    return this.state;
+  }
+
+  /**
+   * Registers a sender.
+   * Sends the full current state as the initial message, and then patches.
+   * This method blocks until the sender is disposed.
+   */
+  async useSender(sender: AwaitableSender<ClusterStoreUpdate>): Promise<void> {
```

Comment:
This API feels somewhat unidiomatic. I keep comparing it in my head to EventEmitter. It's not often I think that we fire off an async function without awaiting it. With EventEmitter this would be solved with events somehow I assume.

We want to "unregister" a sender once it's disposed. I wonder if we should have `sender.onDispose(callback)` instead?

I'll come back to it tomorrow.

Reply:
I think we could try adding just a simple `sender.onDispose` property that could be set within `useSender` instead of `sender. This would avoid that hanging promise of `sender.whenDisposed`.

I left some comments under #59642 about `sender.send` possibly returning a rejected promise. It'd mean that `useSender` has to account for `sender.send` failing.

Reply:
> It seems to me that the main point of confusion with `useSender` is that it returns a promise. It's unclear what are we waiting for.

That's a great observation, I think that's what made this API a bit icky! The new version does feel better.